ykrt/src/compile/jitc_yk/opt/mod.rs

```diff
@@ -782,7 +782,20 @@ impl Opt {
         loop {
             off += pa_inst.off();
             match self.an.op_map(&self.m, pa_inst.ptr(&self.m)) {
-                Operand::Const(_) => todo!(),
+                Operand::Const(cidx) => {
+                    let Const::Ptr(cptr) = self.m.const_(cidx) else {
+                        panic!();
+                    };
+                    // Important details:
+                    //  - offsets can be negative
+                    //  - LLVM semantics require the offset to be sign extended or truncated to
+                    //    the "pointer index type" before computing the offset.
+                    let cidx = self
+                        .m
+                        .insert_const(Const::Ptr((*cptr as isize + off as isize) as usize))?;
```

Comment:
We avoid `as` if at all possible. These should probably be `isize::try_from(...).unwrap()` (etc.) and you will need to decide if you want `saturating_add` or `wrapped_add` etc. to deal with poison cases. `+` will bork in debug mode when it detects overflow, which is not what we want, particularly as these situations do happen because `select` later gets rid of the poisoned value.

Reply:
By `as` casting the offset to isize we truncate or sign extend, as is (I believe) required by LLVM.

This section of the [llvm manual](https://llvm.org/docs/LangRef.html#id238):

> The indices are first converted to offsets in the pointer’s index type. If the currently indexed type is a struct type, the struct offset corresponding to the index is sign-extended or truncated to the pointer index type. Otherwise, the index itself is sign-extended or truncated, and then multiplied by the type allocation size (that is, the size rounded up to the ABI alignment) of the currently indexed type.

I must admit, I find this part of the documentation confusing. I hope I've interpreted it correctly.

Reply:
The problem with `as` is that if you change the types later, it still "works" but does weird things. It's better to use the full type names, as we do elsewhere. Only if you *really* have to should you use `as` (regrettably sometimes you do have to; I don't *think* this is one of those cases, but I might be wrong).

Reply:
BTW, in terms of the `+` stuff, the `BinOp::Add` optimisation probably serves as a useful template for this.

Reply:
Fixed in 9055875dbb7aca0c38f37f40201e87ec2f050ebd with the caveat that in the JIT IR a constant pointer is assumed to be the same size as `usize`, so we have a similar problem as in the last PR.

https://github.com/ykjit/yk/blob/13c330cccbdc750e05d4a032c84a91ed17ee172d/ykrt/src/compile/jitc_yk/jit_ir/mod.rs#L1298

I'm not sure if I should try to tackle this in this PR or not though.

Reply:
Don't tackle `Ptr(usize)` in this PR: there's an element of original sin in the JIT IR. Fixing it won't be a small job.

In terms of the 9055875 (which -- thanks GitHub! -- I can't add comments to), `unsafe transmute` is worse than `as` :p

Looking at the code after the change, the good news is that we *do* have to use `as` but it's for the safe (in our case...) idiom of "bitcast from X to Y where X and Y are the same bitwidth". So we probably just need:

```
.insert_const(Const::Ptr((*ptr as isize).wrapping_add(off)))?
```